receiver/vcenterreceiver/client.go

```diff
@@ -0,0 +1,235 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vcenterreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vcenterreceiver"
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/performance"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/vim25"
+	vt "github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/govmomi/vsan"
+	"github.com/vmware/govmomi/vsan/types"
+)
+
+// vcenterClient is a client that
+type vcenterClient struct {
+	moClient   *govmomi.Client
+	vimDriver  *vim25.Client
+	vsanDriver *vsan.Client
+	finder     *find.Finder
+	pc         *property.Collector
+	cfg        *Config
+}
+
+func newVmwarevcenterClient(c *Config) *vcenterClient {
+	return &vcenterClient{
+		cfg: c,
+	}
+}
+
+// EnsureConnection will establish a connection to the vSphere SDK if not already established
+func (vc *vcenterClient) EnsureConnection(ctx context.Context) error {
+	if vc.moClient == nil {
+		sdkURL, err := vc.cfg.SDKUrl()
+		if err != nil {
+			return err
+		}
+		client, err := govmomi.NewClient(ctx, sdkURL, vc.cfg.MetricsConfig.Insecure)
+		if err != nil {
+			return fmt.Errorf("unable to connect to vSphere SDK on listed endpoint: %w", err)
+		}
+		tlsCfg, err := vc.cfg.MetricsConfig.LoadTLSConfig()
+		if err != nil {
+			return err
+		}
+		client.DefaultTransport().TLSClientConfig = tlsCfg
+		user := url.UserPassword(vc.cfg.MetricsConfig.Username, vc.cfg.MetricsConfig.Password)
+		err = client.Login(ctx, user)
+		if err != nil {
+			return fmt.Errorf("unable to login to vcenter sdk: %w", err)
+		}
+		vc.moClient = client
+		vc.vimDriver = client.Client
+		vc.pc = property.DefaultCollector(vc.vimDriver)
+		vc.finder = find.NewFinder(vc.vimDriver)
+		// swallowing error because not all vCenters need vSAN API enabled
+		_ = vc.connectVSAN(ctx)
```

Comment:
What happens if vsan fails to connect here due to a temporary issue? It will never try again. 

It seems like we keep tripping over vsan, and it's not currently testable anyways. Do you want to remove it from this PR? I think we can merge this much quicker that way. You can always follow up with another PR to add vsan.